### ISSUE TITLE
Warnings being produced when testing optim.

### ISSUE THREAD

**Issue by OliverFM** [pytorch/pytorch#67696]
## 🐛 Bug

This is a small issue: pytorch tests have warnings when run on master (hash: 82f7f8d471241b0f417f932c59558774cc273041).

While this is clearly not a glaring issue, I think that tests ought not to produce warnings when running. It makes it harder for developers to see whether or not that is produced by their work.
Secondly, I think it is generally worth following the paradigms we expect to code with when we write tests.
Does this seem reasonable?

## To Reproduce

1. Install torch.
2. `cd pytorch`
3. `python test/test_optim.py`

Error messages:
```
/home/oliver/code/playground/pytorch-master/torch/optim/asgd.py:34: UserWarning: optimizer contains a parameter group with duplicate parameters; in future, this will cause an error; see github.com/pytorch/pytorch/issues/40967 for more information
  super(ASGD, self).__init__(params, defaults)
 ```
 This occurs for, as far as I can tell, every optimiser. As the error is the same, I will include only one copy.
 
 ```
 ........................../home/oliver/code/playground/pytorch-master/torch/optim/lr_scheduler.py:129: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
```
Is produced several times too. Local testing makes me 99% sure this is caused in this snippet:
```
for _i in range(200):
            for scheduler in schedulers:
                if isinstance(scheduler, ReduceLROnPlateau):
                    val_loss = fn()
                    scheduler.step(val_loss)
                else:
                    scheduler.step()
            optimizer.step(fn)
```
[Source of this test code.](https://github.com/pytorch/pytorch/blob/82f7f8d471241b0f417f932c59558774cc273041/test/test_optim.py#L128)
  

## Expected behavior

Test should pass and issue no warnings.

## Environment

```
Collecting environment information...
PyTorch version: 1.11.0a0+gitcd51d2a
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~18.04) 9.4.0
Clang version: Could not collect # manual note, I built torch with `clang-6.0`
CMake version: version 3.19.6
Libc version: glibc-2.27

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-87-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti
GPU 4: GeForce GTX 1080 Ti
GPU 5: GeForce GTX 1080 Ti
GPU 6: GeForce GTX 1080 Ti
GPU 7: GeForce GTX 1080 Ti

Nvidia driver version: 450.102.04
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.11.0a0+git82f7f8d
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-include               2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.21.2           py38h20f2e39_0
[conda] numpy-base                1.21.2           py38h79a1101_0
[conda] torch                     1.11.0a0+git82f7f8d           dev_0    <develop>
```

## Additional context

One of the two issues is linked to #40967 -- a fix was made to emit a warning, but our tests were not updated.
I am happy to fix all of these issues, if that would be helpful!

cc @vincentqb @jbschlosser @albanD @mruberry

**Comment by albanD** [comment]
Thanks for the issue. It would be great if you could send a PR fixing that:
- If this is actually testing this particular case, then we should catch the warning and expect it to be raised.
- If it is just bad practice in a test that is checking something unrelated, we should fix that test to properly use the API.

**Comment by OliverFM** [comment]
> Thanks for the issue. It would be great if you could send a PR fixing that:
> 
>     * If this is actually testing this particular case, then we should catch the warning and expect it to be raised.
> 
>     * If it is just bad practice in a test that is checking something unrelated, we should fix that test to properly use the API.

Awesome, will do!
I think that this falls under the second category in each case. I will double check when I send out the PR though.

**Comment by OliverFM** [comment]
Regarding point 1, the tests also raise warnings:
```
............../home/oliver/code/playground/pytorch-master/torch/optim/lr_scheduler.py:154: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)
 ```
 which are caught -- but still get logged to the console, any idea on how hard it would be to prevent them being logged when we catch them?

**Comment by albanD** [comment]
If you catch them with `with warnings.catch_warnings` they should not be printed AFAIK
We do that everywhere in the tests.

**Comment by OliverFM** [comment]
Simple question: do I need to configure anything or be granted to push a branch to the repo?
I have made some initial changes, which I tried to push, but have had the results rejected:
```
❯ git push --set-upstream origin oliverfm/fix-test-optim-warnings/issue-67696
ERROR: Permission to pytorch/pytorch.git denied to OliverFM.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

I think that I have set the remote correctly, following [this post](https://stackoverflow.com/questions/14762034/push-to-github-without-a-password-using-ssh-key)

**Comment by albanD** [comment]
You can fork pytorch under your own username and use that one as the upstream. Then you will be able to open a PR from your branch in your fork to pytorch/pytorch master branch.

**Comment by OliverFM** [comment]
Got a partial solution ready for review: #67756

**Comment by OliverFM** [comment]
Next partial fix: #67954

The test is now, fine, we just need to fix the SequentialLR scheduler itself
